File: zig/minsk/src/code_analysis/syntax/Parser.zig
const std = @import("std");
const SyntaxToken = @import("SyntaxToken.zig");
const Lexer = @import("Lexer.zig");
const ExpressionSyntax = @import("ExpressionSyntax.zig");
const BinaryExpressionSyntax = @import("BinaryExpressionSyntax.zig");
const UnaryExpressionSyntax = @import("UnaryExpressionSyntax.zig");
const LiteralExpressionSyntax = @import("LiteralExpressionSyntax.zig");
const ParenthesizedExpressionSyntax = @import("ParenthesizedExpressionSyntax.zig");
const AssignmentExpressionSyntax = @import("AssignmentExpressionSyntax.zig");
const NameExpressionSyntax = @import("NameExpressionSyntax.zig");
const IfStatementSyntax = @import("IfStatementSyntax.zig");
const ElseClauseSyntax = @import("ElseClauseSyntax.zig");
const WhileStatementSyntax = @import("WhileStatementSyntax.zig");

const StatementSyntax = @import("StatementSyntax.zig");
const BlockStatementSyntax = @import("BlockStatementSyntax.zig");
const ExpressionStatementSyntax = @import("ExpressionStatementSyntax.zig");
const VariableDeclarationSyntax = @import("VariableDeclarationSyntax.zig");

const SyntaxKind = @import("syntax_kind.zig").SyntaxKind;
const syntax_facts = @import("syntax_facts.zig");
const CompilationUnitSyntax = @import("CompilationUnitSyntax.zig");
const DiagnosticBag = @import("../DiagnosticBag.zig");
const SourceText = @import("../text/SourceText.zig");

const AllocError = std.mem.Allocator.Error;

allocator: std.mem.Allocator,
source: *const SourceText,
tokens: []SyntaxToken,
position: usize = 0,
diagnostics: DiagnosticBag,

const Self = @This();

pub fn init(allocator: std.mem.Allocator, source: *const SourceText) !Self {
    var lexer = try Lexer.init(allocator, source);
    defer lexer.deinit();
    var tokens = std.ArrayList(SyntaxToken).init(allocator);
    defer tokens.deinit();

    while (try lexer.lex()) |token| {
        if (token.kind == .whitespace_token or token.kind == .bad_token)
            continue;
        try tokens.append(token);
    }
    var diagnostics = DiagnosticBag.init(allocator);
    try diagnostics.extend(&lexer.diagnostics);

    return .{
        .allocator = allocator,
        .source = source,
        .tokens = try tokens.toOwnedSlice(),
        .diagnostics = diagnostics,
    };
}

pub fn deinit(self: Self) void {
    self.allocator.free(self.tokens);
    self.diagnostics.deinit();
}

fn peek(self: Self, offset: usize) SyntaxToken {
    const index = self.position + offset;
    if (index >= self.tokens.len) {
        return self.tokens[self.tokens.len - 1];
    }
    return self.tokens[index];
}

fn current(self: Self) SyntaxToken {
    return self.peek(0);
}

fn nextToken(self: *Self) SyntaxToken {
    const result = self.current();
    self.position += 1;
    return result;
}

fn matchToken(self: *Self, kind: SyntaxKind) AllocError!SyntaxToken {
    if (self.current().kind == kind) {
        return self.nextToken();
    }

    try self.diagnostics.reportUnexpectedToken(
        self.current().span(),
        self.current().kind,
        kind,
    );
    return SyntaxToken.init(
        kind,
        self.current().position,
        "",
        null,
    );
}

pub fn parseCompilationUnit(self: *Self) AllocError!*CompilationUnitSyntax {
    const statement = try self.parseStatement();
    const end_of_file_token = try self.matchToken(.end_of_file_token);
    return try CompilationUnitSyntax.init(
        self.allocator,
        statement,
        end_of_file_token,
    );
}

fn takeDiagnostics(self: *Self) DiagnosticBag {
    const result = self.diagnostics;
    self.diagnostics = DiagnosticBag.init(self.allocator);
    return result;
}

fn parseStatement(self: *Self) AllocError!*StatementSyntax {
    return switch (self.current().kind) {
        .open_brace_token => try self.parseBlockStatement(),
        .let_keyword, .var_keyword => try self.parseVariableDeclaration(),
        .if_keyword => try self.parseIfStatement(),
        .while_keyword => try self.parseWhileStatement(),
        else => try self.parseExpressionStatement(),
    };
}

fn parseBlockStatement(self: *Self) AllocError!*StatementSyntax {
    var statements = std.ArrayList(*StatementSyntax).init(self.allocator);

    const open_brace_token = try self.matchToken(.open_brace_token);
    while (self.current().kind != .end_of_file_token and
        self.current().kind != .close_brace_token)
    {
        const statement = try self.parseStatement();
        try statements.append(statement);
    }
    const close_brace_token = try self.matchToken(.close_brace_token);
    return try BlockStatementSyntax.init(
        self.allocator,
        open_brace_token,
        try statements.toOwnedSlice(),
        close_brace_token,
    );
}

fn parseIfStatement(self: *Self) AllocError!*StatementSyntax {
    const keyword_token = try self.matchToken(.if_keyword);
    const condition = try self.parseExpression();
    const then_statement = try self.parseStatement();
    const else_clause = try self.parseOptionalElseClause();
    return try IfStatementSyntax.init(
        self.allocator,
        keyword_token,
        condition,
        then_statement,
        else_clause,
    );
}

fn parseWhileStatement(self: *Self) AllocError!*StatementSyntax {
    const keyword_token = try self.matchToken(.while_keyword);
    const condition = try self.parseExpression();
    const body = try self.parseStatement();
    return try WhileStatementSyntax.init(
        self.allocator,
        keyword_token,
        condition,
        body,
    );
}

fn parseOptionalElseClause(self: *Self) AllocError!?*ElseClauseSyntax {
    if (self.current().kind != .else_keyword)
        return null;

    const keyword_token = self.nextToken();
    const statement = try self.parseStatement();
    return try ElseClauseSyntax.init(self.allocator, keyword_token, statement);
}

fn parseVariableDeclaration(self: *Self) AllocError!*StatementSyntax {
    const expected: SyntaxKind = switch (self.current().kind) {
        .let_keyword => .let_keyword,
        else => .var_keyword,
    };
    const keyword_token = try self.matchToken(expected);
    const identifier_token = try self.matchToken(.identifier_token);
    const equals_token = try self.matchToken(.equals_token);
    const initializer = try self.parseExpression();
    return try VariableDeclarationSyntax.init(
        self.allocator,
        keyword_token,
        identifier_token,
        equals_token,
        initializer,
    );
}

fn parseExpressionStatement(self: *Self) AllocError!*StatementSyntax {
    const expression = try self.parseExpression();
    return try ExpressionStatementSyntax.init(self.allocator, expression);
}

fn parseExpression(self: *Self) AllocError!*ExpressionSyntax {
    return try self.parseAssignmentExpression();
}

fn parseAssignmentExpression(self: *Self) !*ExpressionSyntax {
    if (self.peek(0).kind == .identifier_token and
        self.peek(1).kind == .equals_token)
    {
        const identifier_token = self.nextToken();
        const equals_token = self.nextToken();
        const expression = try self.parseAssignmentExpression();
        return try AssignmentExpressionSyntax.init(self.allocator, identifier_token, equals_token, expression);
    }

    return try self.parseBinaryExpression(0);
}

fn parseBinaryExpression(self: *Self, parent_precedence: usize) !*ExpressionSyntax {
    const unary_operator_precedence = syntax_facts.unaryOperatorPrecedence(self.current().kind);
    var left = if (unary_operator_precedence != 0 and unary_operator_precedence >= parent_precedence) blk: {
        const operator_token = self.nextToken();
        const operand = try self.parseBinaryExpression(unary_operator_precedence);
        break :blk try UnaryExpressionSyntax.init(self.allocator, operator_token, operand);
    } else try self.parsePrimaryExpression();

    errdefer left.deinit(self.allocator);

    while (true) {
        const precedence = syntax_facts.binaryOperatorPrecedence(self.current().kind);
        if (precedence == 0 or precedence <= parent_precedence)
            break;

        const operator_token = self.nextToken();
        const right = try self.parseBinaryExpression(precedence);
        errdefer right.deinit(self.allocator);
        left = try BinaryExpressionSyntax.init(self.allocator, left, operator_token, right);
    }

    return left;
}

fn parsePrimaryExpression(self: *Self) !*ExpressionSyntax {
    return switch (self.current().kind) {
        .open_parenthesis_token => try self.parseParenthesizedExpression(),
        .false_keyword, .true_keyword => try self.parseBooleanLiteral(),
        .number_token => try self.parseNumberLiteral(),
        else => try self.parseNameExpression(),
    };
}

fn parseParenthesizedExpression(self: *Self) !*ExpressionSyntax {
    const left = try self.matchToken(.open_parenthesis_token);
    const expression = try self.parseExpression();
    errdefer expression.deinit(self.allocator);
    const right = try self.matchToken(.close_parenthesis_token);
    return try ParenthesizedExpressionSyntax.init(
        self.allocator,
        left,
        expression,
        right,
    );
}

fn parseBooleanLiteral(self: *Self) !*ExpressionSyntax {
    const value = self.current().kind == .true_keyword;
    const keyword_token = if (value)
        self.nextToken()
    else
        try self.matchToken(.false_keyword);
    return try LiteralExpressionSyntax.init(
        self.allocator,
        keyword_token,
        .{ .boolean = value },
    );
}

fn parseNameExpression(self: *Self) !*ExpressionSyntax {
    const identifier_token = try self.matchToken(.identifier_token);
    return try NameExpressionSyntax.init(self.allocator, identifier_token);
}

fn parseNumberLiteral(self: *Self) !*ExpressionSyntax {
    const number_token = try self.matchToken(.number_token);
    return try LiteralExpressionSyntax.init(
        self.allocator,
        number_token,
        number_token.value orelse .{ .integer = 0 },
    );
}

File: ocaml/lib/code_analysis/syntax/parser.ml
open Node.Expression
open Token
open Facts

type t = {
  tokens : Token.t array;
  diagnostics : Diagnostic_bag.t;
  mutable position : int;
}

let make text =
  let tokens, diagnostics = Lexer.lex text in
  let tokens =
    Array.to_seq tokens
    |> Seq.filter (fun (t : Token.t) ->
           match t.kind with Bad | Whitespace -> false | _ -> true)
    |> Array.of_seq
  in
  { tokens; diagnostics; position = 0 }

let peek offset p =
  let i = p.position + offset in
  if i >= Array.length p.tokens then p.tokens.(Array.length p.tokens - 1)
  else p.tokens.(i)

let current = peek 0

let next_token p =
  let c = current p in
  p.position <- p.position + 1;
  c

let match_fail actual_kind expected_kind p =
  Diagnostic_bag.report_unexpected_token
    (current p |> Token.span)
    ~actual_kind ~expected_kind p.diagnostics;
  Token.make expected_kind (current p).position "" None

let match_token kind p =
  match (current p).kind with
  | k when k = kind -> next_token p
  | k -> match_fail k kind p

let rec parse_expression p = parse_assignment_expression p

and parse_assignment_expression p =
  if (peek 0 p).kind = Identifier && (peek 1 p).kind = Equals then
    let identifier_token = next_token p in
    let operator_token = next_token p in
    let right = parse_assignment_expression p in
    Assignment (Assignment.make identifier_token operator_token right)
  else parse_binary_expression ~parent_precedence:0 p

and parse_binary_expression ~parent_precedence p =
  let rec next left =
    let precedence = binary_operator_precedence (current p).kind in
    if precedence = 0 || precedence <= parent_precedence then left
    else
      let operator_token = next_token p in
      let right = parse_binary_expression ~parent_precedence:precedence p in
      Binary (Binary.make left operator_token right) |> next
  in
  let unary_precedence = unary_operator_precedence (current p).kind in
  (if unary_precedence = 0 || unary_precedence < parent_precedence then
     parse_primary_expression p
   else
     let operator_token = next_token p in
     let operand =
       parse_binary_expression ~parent_precedence:unary_precedence p
     in
     Unary (Unary.make operator_token operand))
  |> next

and parse_primary_expression p =
  match (current p).kind with
  | OpenParenthesis -> parse_parenthesized_expression p
  | FalseKeyword | TrueKeyword -> parse_boolean_literal p
  | Number -> parse_number_literal p
  | _ -> parse_name_expression p

and parse_parenthesized_expression p =
  let left = match_token OpenParenthesis p in
  let expression = parse_expression p in
  let right = match_token CloseParenthesis p in
  Parenthesized (Parenthesized.make left expression right)

and parse_boolean_literal p =
  let is_true = (current p).kind = TrueKeyword in
  let keyword_token =
    match_token (if is_true then TrueKeyword else FalseKeyword) p
  in
  Literal (Literal.make keyword_token (Some (Bool is_true)))

and parse_number_literal p =
  let number_token = match_token Number p in
  Literal (Literal.make number_token number_token.value)

and parse_name_expression p =
  let identifier_token = match_token Identifier p in
  Name (Name.make identifier_token)

let parse text =
  let p = make text in
  let expr = parse_expression p in
  let end_of_file_token = match_token EndOfFile p in
  (p.diagnostics |> Diagnostic_bag.to_array, expr, end_of_file_token)
